### PR TITLE
feat: add sample CRV and UNI strats for CREAM

### PR DESCRIPTION
**Pull request by storming0x**
Change Log:
* adds CRV Cream Strat sample
* adds UNI Cream Strat sample
* adds integration tests for strats/vaults use cases deposit/withdraw/withdrawAll/earn/harvest

### CODE REVIEW

File: contracts/strategies/StrategyCreamCRV.sol [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+        uint256 _withdrew = _after.sub(_before);
+        return _withdrew;
+    }
+
+    function balanceOfWant() public view returns (uint256) {
+        return IERC20(want).balanceOf(address(this));
+    }
+
+    function _withdrawC(uint256 amount) internal {
+        cToken(crCRV).redeem(amount);
```
Comment (andrecronje) [inline contracts/strategies/StrategyCreamCRV.sol]: Redeem doesn't revert with an error code if insufficient funds, require(success...

Comment (storming0x) [reply contracts/strategies/StrategyCreamCRV.sol]: added the require check on latest on commit 018df91, tested against specific amounts but still need more testing on `vault.withdrawAll` to confirm it works for all cases.

Comment (storming0x) [reply contracts/strategies/StrategyCreamCRV.sol]: latest tests adds testing for withdraws from vault and from strategy

File: contracts/strategies/StrategyCreamCRV.sol [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    function _withdrawC(uint256 amount) internal {
+        // 0=success else fails with error code
+        require(cToken(crCRV).redeem(amount) == 0, "cToken redeem failed!");
+    }
+
+    function balanceCInToken() public view returns (uint256) {
+        // Mantisa 1e18 to decimals
+        uint256 b = balanceC();
+        if (b > 0) {
+            b = b.mul(cToken(crCRV).exchangeRateStored()).div(1e18);
```
Comment (fubuloubu) [inline contracts/strategies/StrategyCreamCRV.sol:172]: is `cToken(crCRV).exchangeRateStored()` able to manipulated?

File: contracts/strategies/StrategyCreamUNI.sol [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    function _withdrawC(uint256 amount) internal {
+        // 0=success else fails with error code
+        require(cToken(crUNI).redeem(amount) == 0, "cToken redeem failed!");
+    }
+
+    function balanceCInToken() public view returns (uint256) {
+        // Mantisa 1e18 to decimals
+        uint256 b = balanceC();
+        if (b > 0) {
+            b = b.mul(cToken(crUNI).exchangeRateStored()).div(1e18);
```
Comment (fubuloubu) [inline contracts/strategies/StrategyCreamUNI.sol:172]: Is `cToken(crUNI).exchangeRateStored()` able to be manipulated?

File: tests/functional/strategies/test_strategy_uni_cream.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+    # expectations
+    assert token.balanceOf(user) == Wei("100 ether")
+
+
+def test_vault_withdraw_with_fee(strategy, vault, token, cToken, gov, rewards):
+    # setup accounts and expectations
+    uni_holder = accounts.at(UNI_HOLDER, force=True)
+    user = accounts[9]
+    # fund user wallet with CRV
```
Comment (lbertenasco) [inline tests/functional/strategies/test_strategy_uni_cream.py]:
```suggestion
    # fund user wallet with UNI
```
typo

Comment (storming0x) [reply tests/functional/strategies/test_strategy_uni_cream.py]: fixed in latest commit

File: contracts/strategies/StrategyCreamCRV.sol (hunk 2 above)
Comment (storming0x) [reply contracts/strategies/StrategyCreamCRV.sol:172]: This is the recommended way to calculate the balance in underlying token from the cToken by compound and applies also to CREAM as well. Was used and taken from deployed CreamYFI strat here:  https://github.com/iearn-finance/yearn-protocol/blob/1c6e7c19d2ee223894a4ff7f90a3842d454b5194/contracts/strategies/StrategyCreamYFI.sol#L169

My understanding from Compound's docs is that the base rate is based on market for token and would not change or be subject to manipulation in a single block.
https://compound.finance/docs/ctokens#introduction
@andrecronje can confirm

File: contracts/strategies/StrategyCreamUNI.sol (hunk 3 above)
Comment (storming0x) [reply contracts/strategies/StrategyCreamUNI.sol:172]: answered above on first comment

File: tests/functional/strategies/test_strategy_crv_cream.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+@pytest.fixture
+def vault(gov, controller, strategy, token):
+    # vault setup
+    vault = gov.deploy(yVault, token, controller)
+    vault.setMin(10000)
+
+    # enable strat
+    controller.approveStrategy(token, strategy, {"from": gov})
+    controller.setStrategy(token, strategy, {"from": gov})
+    controller.setVault(token, vault)
```
Comment (storming0x) [inline tests/functional/strategies/test_strategy_crv_cream.py:50]: set vault

**Comment by storming0x** [comment]
migrated to v2 so closing this PR

**Pull request closed without merging**